packages/angular/ssr/node/src/response.ts

```diff
@@ -52,27 +53,49 @@ export async function writeResponseToNodeResponse(
 
   try {
     const reader = body.getReader();
-
-    destination.on('close', () => {
-      reader.cancel().catch((error) => {
+    function cancelStream(error?: Error) {
+      reader.cancel(error).catch(() => {
         // eslint-disable-next-line no-console
         console.error(
           `An error occurred while writing the response body for: ${destination.req.url}.`,
           error,
         );
       });
-    });
-
-    // eslint-disable-next-line no-constant-condition
-    while (true) {
-      const { done, value } = await reader.read();
-      if (done) {
-        destination.end();
-        break;
+      if (error) {
+        destination.destroy(error);
       }
+    }
 
-      (destination as ServerResponse).write(value);
+    function handleStream({
```

Comment:
The new logic seems a bit hard to follow. In my opinion, it could be simplified significantly by using `async/await` instead of relying on callbacks. Additionally, the use of `handleStream` could be avoided, which would make the flow cleaner and easier to understand.


```ts
destination.on('close', handleClose);
destination.on('error', handleError);

try {
  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      destination.end();
      break;
    }

    if (!(destination as ServerResponse).write(value)) {
      await new Promise<void>((resolve) => destination.once('drain', resolve));
    }
  }
} catch (error) {
  handleError(error);
} finally {
  handleClose();
}
```